File: JavaScript/Tests/math-tests.js
var degreesToRadians = adhan.Math.degreesToRadians;
var radiansToDegrees = adhan.Math.radiansToDegrees;
var normalizeWithBound = adhan.Math.normalizeWithBound;
var unwindAngle = adhan.Math.unwindAngle;
var closestAngle = adhan.Math.closestAngle;
var timeComponents = adhan.Math.timeComponents;
var roundedMinute = adhan.Date.roundedMinute;
//
// Math tests
//

QUnit.test("Angle Conversion", function(assert) {
	assert.equal(radiansToDegrees(Math.PI), 180);
	assert.equal(degreesToRadians(180), Math.PI);
	assert.equal(radiansToDegrees(Math.PI/2), 90);
	assert.equal(degreesToRadians(90), Math.PI/2);
});

QUnit.test("Normalizing", function(assert) {
	assert.equal(normalizeWithBound(2.0, -5), -3);
	assert.equal(normalizeWithBound(-4, -5), -4);
	assert.equal(normalizeWithBound(-6, -5), -1);

	assert.equal(normalizeWithBound(-1, 24), 23);
	assert.equal(normalizeWithBound(1, 24), 1);
	assert.equal(normalizeWithBound(49.0, 24), 1);

	assert.equal(normalizeWithBound(361, 360), 1);
	assert.equal(normalizeWithBound(360, 360), 0);
	assert.equal(normalizeWithBound(259, 360), 259);
	assert.equal(normalizeWithBound(2592.0, 360), 72);

	assert.equal(unwindAngle(-45), 315);
	assert.equal(unwindAngle(361.0), 1);
	assert.equal(unwindAngle(360), 0);
	assert.equal(unwindAngle(259.0), 259);
	assert.equal(unwindAngle(2592), 72);

	QUnit.close(normalizeWithBound(360.1, 360), 0.1, 0.01);
});

QUnit.test("Closest Angle", function(assert) {
    assert.equal(closestAngle(360.0), 0);
    assert.equal(closestAngle(361.0), 1);
    assert.equal(closestAngle(1.0), 1);
    assert.equal(closestAngle(-1.0), -1);
    assert.equal(closestAngle(-181.0), 179);
    assert.equal(closestAngle(180.0), 180);
    assert.equal(closestAngle(359.0), -1);
    assert.equal(closestAngle(-359.0), 1);
    assert.equal(closestAngle(1261.0), -179);
    QUnit.close(closestAngle(-360.1), -0.1, 0.01);
});

QUnit.test("Time Components", function(assert) {
    var comps1 = timeComponents(15.199)
    assert.equal(comps1.hours, 15);
    assert.equal(comps1.minutes, 11);
    assert.equal(comps1.seconds, 56);

    var comps2 = timeComponents(1.0084);
    assert.equal(comps2.hours, 1);
    assert.equal(comps2.minutes, 0);
    assert.equal(comps2.seconds, 30);

    var comps3 = timeComponents(1.0083);
    assert.equal(comps3.hours, 1);
    assert.equal(comps3.minutes, 0);

    var comps4 = timeComponents(2.1);
    assert.equal(comps4.hours, 2);
    assert.equal(comps4.minutes, 6);

    var comps5 = timeComponents(3.5);
    assert.equal(comps5.hours, 3);
    assert.equal(comps5.minutes, 30);
});

QUnit.test("Minute Rounding", function(assert) {
	var date1 = roundedMinute(new Date(2015, 0, 1, 10, 2, 29));
	assert.equal(date1.getMinutes(), 2);
	assert.equal(date1.getSeconds(), 0);

	var date2 = roundedMinute(new Date(2015, 0, 1, 10, 2, 31));
	assert.equal(date2.getMinutes(), 3);
	assert.equal(date2.getSeconds(), 0);
});

File: Shared/Swift/AdhanTests/MathTests.swift
//
//  MathTests.swift
//  Adhan
//
//  Created by Ameir Al-Zoubi on 2/21/16.
//  Copyright © 2016 Batoul Apps. All rights reserved.
//

import XCTest
@testable import Adhan

class MathTests: XCTestCase {
    
    func testAngleConversion() {
        XCTAssertEqual(Double.pi.radiansToDegrees(), 180.0)
        XCTAssertEqual(180.0.degreesToRadians(), Double.pi)
        XCTAssertEqual((Double.pi / 2).radiansToDegrees(), 90.0)
        XCTAssertEqual(90.0.degreesToRadians(), (Double.pi / 2))
    }
    
    func testNormalizing() {
        XCTAssertEqual(2.0.normalizeWithBound(max: -5), -3)
        XCTAssertEqual((-4.0).normalizeWithBound(max: -5), -4)
        XCTAssertEqual((-6.0).normalizeWithBound(max: -5), -1)
        
        XCTAssertEqual((-1.0).normalizeWithBound(max: 24), 23)
        XCTAssertEqual(1.0.normalizeWithBound(max: 24), 1)
        XCTAssertEqual(49.0.normalizeWithBound(max: 24), 1)
        
        XCTAssertEqual(361.0.normalizeWithBound(max: 360), 1)
        XCTAssertEqual(360.0.normalizeWithBound(max: 360), 0)
        XCTAssertEqual(259.0.normalizeWithBound(max: 360), 259)
        XCTAssertEqual(2592.0.normalizeWithBound(max: 360), 72)
        
        XCTAssertEqual((-45.0).unwindAngle(), 315)
        XCTAssertEqual(361.0.unwindAngle(), 1)
        XCTAssertEqual(360.0.unwindAngle(), 0)
        XCTAssertEqual(259.0.unwindAngle(), 259)
        XCTAssertEqual(2592.0.unwindAngle(), 72)
        
        XCTAssertEqual(360.1.normalizeWithBound(max: 360), 0.1, accuracy: 0.01)
    }
    
    func testClosestAngle() {
        XCTAssertEqual(360.0.closestAngle(), 0)
        XCTAssertEqual(361.0.closestAngle(), 1)
        XCTAssertEqual(1.0.closestAngle(), 1)
        XCTAssertEqual((-1.0).closestAngle(), -1)
        XCTAssertEqual((-181.0).closestAngle(), 179)
        XCTAssertEqual(180.0.closestAngle(), 180)
        XCTAssertEqual(359.0.closestAngle(), -1)
        XCTAssertEqual((-359.0).closestAngle(), 1)
        XCTAssertEqual(1261.0.closestAngle(), -179)
        XCTAssertEqual((-360.1).closestAngle(), -0.1, accuracy: 0.01)
    }
    
    func testTimeComponents() {
        let comps1 = 15.199.timeComponents()!
        XCTAssertEqual(comps1.hours, 15)
        XCTAssertEqual(comps1.minutes, 11)
        XCTAssertEqual(comps1.seconds, 56)
        
        let comps2 = 1.0084.timeComponents()!
        XCTAssertEqual(comps2.hours, 1)
        XCTAssertEqual(comps2.minutes, 0)
        XCTAssertEqual(comps2.seconds, 30)
        
        let comps3 = 1.0083.timeComponents()!
        XCTAssertEqual(comps3.hours, 1)
        XCTAssertEqual(comps3.minutes, 0)
        
        let comps4 = 2.1.timeComponents()!
        XCTAssertEqual(comps4.hours, 2)
        XCTAssertEqual(comps4.minutes, 6)
        
        let comps5 = 3.5.timeComponents()!
        XCTAssertEqual(comps5.hours, 3)
        XCTAssertEqual(comps5.minutes, 30)
    }
    
    func testMinuteRounding() {
        let cal = Calendar(identifier: Calendar.Identifier.gregorian)
        
        let comps1 = DateComponents(year: 2015, month: 1, day: 1, hour: 10, minute: 2, second: 29)
        let date1 = cal.date(from: comps1)!.roundedMinute()
        XCTAssertEqual(cal.component(.minute, from: date1), 2)
        XCTAssertEqual(cal.component(.second, from: date1), 0)
        
        let comps2 = DateComponents(year: 2015, month: 1, day: 1, hour: 10, minute: 2, second: 31)
        let date2 = cal.date(from: comps2)!.roundedMinute()
        XCTAssertEqual(cal.component(.minute, from: date2), 3)
        XCTAssertEqual(cal.component(.second, from: date2), 0)
    }
}
